### PR TITLE
[ENHANCEMENT] Improve ML Model Versioning and Safe Rollbacks

### LINKED ISSUE #1326: [ENHANCEMENT] Improve ML Model Versioning and Safe Rollbacks
## ⚙️ Current Behavior

Machine learning models are currently updated or replaced without a clearly defined versioning or rollback mechanism.
Once a new model is deployed, there is no easy way to track model versions or safely revert to a previous stable model if issues arise.

## 🚀 Proposed Improvement

Introduce a structured model versioning and rollback strategy for all ML models used in the application.
Each model release should have a unique version identifier (e.g., semantic versioning or date-based tags) and support switching between versions without disrupting production.

## 🔍 Why It’s Needed

Without proper model version control, deploying new models can lead to unpredictable behavior, inconsistent predictions, and production instability.
A versioning and rollback mechanism ensures safer deployments, easier debugging, and faster recovery when a model performs poorly or introduces errors.

## 🧩 Possible Implementation

Store models using versioned directories or tags (e.g., model_v1, model_v2, model_2026_02_01).
Maintain a configuration file or environment variable to define the active model version
Log the model version used for each prediction request.
Add a rollback mechanism to quickly switch to a previous stable model if required.
(Optional) Maintain a changelog describing improvements or fixes for each model version.

## ✅ Checklist

- [✅] I’ve reviewed existing issues to ensure this isn’t a duplicate.  
- [✅] I’ve explained how this improves performance or readability.

### PR DESCRIPTION
**Pull request by Aditya8369**
### **Core System (2,000+ lines of code)**

| Component | File | Features |
|-----------|------|----------|
| **Model Manager** | model_manager.py | Version creation, loading, rollback, metadata management |
| **Model Registry** | model_registry.py | Model catalog, configuration, version history |
| **REST API** | model_versioning.py | 8+ endpoints for all operations |
| **Migration Tool** | migrate_models.py | Automated migration from legacy models |
| **Tests** | test_versioning.py | 15+ test cases covering all features |

### **📚 Documentation (2,000+ lines)**

- **README.md** - Complete user guide with best practices
- **CHANGELOG.md** - Version history and performance tracking
- **QUICK_START_MODEL_VERSIONING.md** - Quick reference guide
- **ML_MODEL_VERSIONING_IMPLEMENTATION.md** - Implementation details

## 🎯 Key Features Implemented

✅ **Semantic Versioning** (v1.0.0, v1.1.0, v2.0.0)  
✅ **Date-based Tagging** (2026_02_15 format)  
✅ **Safe Rollback Mechanism** with validation  
✅ **Artifact Management** (encoders, scalers, etc.)  
✅ **Performance Metrics Tracking** per version  
✅ **Complete Audit Logging** (JSON format)  
✅ **Multi-Format Support** (sklearn, Keras, PyTorch, pickle)  
✅ **REST API** with 8 endpoints  
✅ **Automatic Migration** from legacy models  
✅ **Model Health Checks**

## 🏗️ Directory Structure

```
backend/ml_models/
├── model_manager.py          # Core versioning engine
├── model_registry.py         # Model registry
├── migrate_models.py         # Migration script
├── test_versioning.py        # Unit tests
├── README.md                 # Full documentation
├── CHANGELOG.md              # Version history
├── crop_recommendation/      # Versioned models
│   ├── v1.0.0/, v1.1.0/, v1.2.0/
│   └── active_version.json
├── disease_detection/
└── versioning_logs/          # Audit trail
```

### Python API
```python
from backend.ml_models import get_model_manager, ModelType

manager = get_model_manager()

# Create version
manager.create_version('crop_recommendation', model, 'v1.2.0', 
                      ModelType.SKLEARN, artifacts={'encoder': enc})

# Load active model
model = manager.load_model('crop_recommendation')

# Rollback safely
manager.rollback_to_version('crop_recommendation', 'v1.1.0')
```

### REST API
```bash
# List versions
curl http://localhost:5000/api/v1/models/crop_recommendation/versions

# Get active version
curl http://localhost:5000/api/v1/models/crop_recommendation/active

# Rollback
curl -X POST http://localhost:5000/api/v1/models/crop_recommendation/rollback \
  -H "Content-Type: application/json" \
  -d '{"target_version": "v1.1.0"}'
```

## 🚀 Getting Started

1. **Migrate existing models:**
   ```bash
   python backend/ml_models/migrate_models.py --migrate-all
   ```

2. **Run tests:**
   ```bash
   python -m pytest backend/ml_models/test_versioning.py -v
   ```

3. **Start using in your code** (already integrated with crop recommendation routes)

## 📄 Files Modified

- routes.py - Updated to use versioning system with fallback
- app.py - Added model versioning API blueprint

## 🎓 Pre-configured Models

The system comes with 4 models ready to use:
- **crop_recommendation** (v1.2.0) - Active
- **disease_detection** (v1.1.0) - Active  
- **yield_prediction** (v1.0.0) - Active
- **pest_detection** - Ready for future use

All files are fully documented with docstrings, examples, and best practices. The system is **production-ready** with comprehensive error handling, logging, and validation.

closes #1326

### CODE REVIEW

**Comment by Aditya8369** [comment]
@omroy07 consider my PRs under NSoC and add its relevant labels to it
Thank you
